src/core/lib/security/transport/security_connector.c

```diff
@@ -511,14 +545,109 @@ static void ssl_channel_add_handshakers(grpc_exec_ctx *exec_ctx,
           exec_ctx, tsi_create_adapter_handshaker(tsi_hs), &sc->base));
 }
 
+static const char **fill_alpn_protocol_strings(size_t *num_alpn_protocols) {
+  size_t i;
+  const char **alpn_protocol_strings;
+
+  GPR_ASSERT(num_alpn_protocols != NULL);
+  *num_alpn_protocols = grpc_chttp2_num_alpn_versions();
+  alpn_protocol_strings =
+      gpr_malloc(sizeof(const char *) * (*num_alpn_protocols));
+  for (i = 0; i < *num_alpn_protocols; i++) {
+    alpn_protocol_strings[i] = grpc_chttp2_get_alpn_version_index(i);
+  }
+  return alpn_protocol_strings;
+}
+
+/* Attempts to replace the server_handshaker_factory with a new factory using
+ * the provided grpc_server_credentials. Should the credentials not be of type
+ * SSL or should the new factory creation fail, the existing factory will not be
+ * replaced. Returns true on success (new factory created). */
+static bool try_replace_server_handshaker_factory(
+    grpc_ssl_server_security_connector *sc,
+    grpc_ssl_server_certificate_config *config) {
+  size_t num_alpn_protocols = 0;
+  const char **alpn_protocol_strings = NULL;
+  tsi_ssl_server_handshaker_factory *new_handshaker_factory = NULL;
+  tsi_ssl_pem_key_cert_pair *cert_pairs = NULL;
+  tsi_result result;
+
+  if (config == NULL) {
+    gpr_log(GPR_ERROR,
+            "Server certificate config callback returned invalid (NULL) "
+            "config.");
+    return false;
+  }
+
+  gpr_log(GPR_DEBUG, "Using new server certificate config (%p).", config);
+
+  alpn_protocol_strings = fill_alpn_protocol_strings(&num_alpn_protocols);
+  cert_pairs = convert_grpc_to_tsi_cert_pairs(config->pem_key_cert_pairs,
+                                              config->num_key_cert_pairs);
+  result = tsi_create_ssl_server_handshaker_factory_ex(
+      cert_pairs, config->num_key_cert_pairs, config->pem_root_certs,
+      get_tsi_client_certificate_request_type(sc->client_certificate_request),
+      ssl_cipher_suites(), alpn_protocol_strings, (uint16_t)num_alpn_protocols,
+      &new_handshaker_factory);
+  gpr_free(cert_pairs);
+  gpr_free((void *)alpn_protocol_strings);
+
+  if (result != TSI_OK) {
+    gpr_log(GPR_ERROR, "Handshaker factory creation failed with %s.",
+            tsi_result_to_string(result));
+    return false;
+  }
+
+  tsi_ssl_server_handshaker_factory_unref(sc->server_handshaker_factory);
+  sc->server_handshaker_factory = new_handshaker_factory;
+  return true;
+}
+
+/* Attempts to fetch the server credentials if a callback is available. Current
+ * credentials will continue to be used if the callback returns an error.
+ * Returns true if new credentials were sucessfully loaded. */
+static bool try_fetch_ssl_server_credentials(
```

Comment:
`try_fetch_` can be a little misleading, because after fetching, if the credentials are new, it will also try to replace/update...

Reply:
Hmm, I'm open to suggestions. I did consider breaking this function into two, but couldn't find a satisfactory way.

Reply:
`try_reload_`?